### PR TITLE
Add support for a TAP formatter

### LINKED ISSUE #3889: Add TAP formatter
> What is the problem you're trying to solve?

We run stylelint under Jenkins.  Our tooling, including other applications such as ESLint, generally output as TAP, which Jenkins picks up and gives us a pretty output for.  Unlike ESLint, Stylelint doesn't seem to currently have a method of outputting its results to TAP.
> What solution would you like to see?

It would be extremely helpful to have a formatter which outputs the test results as TAP.

### PR DESCRIPTION
**Pull request by clinejj**
This PR adds support for a TAP formatter using an update to the code snippet from @danielsamuels discussed in #3889.

> Which issue, if any, is this issue related to?

Closes #3889

> Is there anything in the PR that needs further explanation?

No, it is self explanatory. Used #3524 as a starting point.

### CODE REVIEW

File: lib/formatters/tapFormatter.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+						`     ruleId: ${warning.rule}`,
+					],
+				);
+			});
+
+			lines.push('---');
+		}
+	});
+
+	lines.push('');
```
Comment (clinejj) [inline lib/formatters/tapFormatter.js:36]: It's not clear to me why this needs to exist, but with some manual testing in another project if this is removed the last result will not be reported for larger sets of results. Would definitely be interested if there's a better approach based on what happens downstream of the formatter.

File: types/stylelint/index.d.ts [hunk 2]
```diff
@@ -91,7 +91,14 @@ declare module 'stylelint' {
 		returnValue?: StylelintStandaloneReturnValue,
 	) => string;
 
-	export type FormatterIdentifier = 'compact' | 'json' | 'string' | 'unix' | 'verbose' | Formatter;
+	export type FormatterIdentifier =
```
Comment (clinejj) [inline types/stylelint/index.d.ts:94]: Added tap as an option but format was adjusted by prettier

File: lib/formatters/tapFormatter.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+			`${result.errored ? 'not ok' : 'ok'} ${index + 1} - ${result.ignored ? 'ignored ' : ''}${
+				result.source
+			}`,
+		);
+
+		if (result.warnings.length > 0) {
+			lines.push(...['---', 'messages:']);
+
+			result.warnings.forEach((warning) => {
+				lines.push(
```
Comment (vankop) [inline lib/formatters/tapFormatter.js:20]: you can use `N` arguments to push in Array.prototype.push https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/push

File: lib/formatters/tapFormatter.js (hunk 1 above)
Comment (vankop) [reply lib/formatters/tapFormatter.js:36]: maybe for some tools `\n` is required in the end

**Review by vankop — APPROVED** [review]
Thanks for your PR

File: lib/formatters/tapFormatter.js (hunk 3 above)
Comment (clinejj) [reply lib/formatters/tapFormatter.js:20]: Great point! Updated

File: lib/formatters/tapFormatter.js (hunk 1 above)
Comment (clinejj) [reply lib/formatters/tapFormatter.js:34]: That makes sense, leaving as-is

**Review by jeddy3 — APPROVED** [review]
Thank you!

**Comment by jeddy3** [comment]
Changelog entry:
> - Added: TAP formatter ([#5062](https://github.com/stylelint/stylelint/pull/5062)).

File: lib/formatters/tapFormatter.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+					` - message: "${warning.text}"`,
+					`   severity: ${warning.severity}`,
+					`   data:`,
+					`     line: ${warning.line}`,
+					`     column: ${warning.column}`,
+					`     ruleId: ${warning.rule}`,
+				);
+			});
+
+			lines.push('---');
```
Comment (Mouvedia) [inline lib/formatters/tapFormatter.js:30]: @ybiquitous

According to https://testanything.org/tap-version-14-specification.html#yaml-diagnostics it should be `...`, right?

File: lib/formatters/tapFormatter.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+	results.forEach((result, index) => {
+		lines.push(
+			`${result.errored ? 'not ok' : 'ok'} ${index + 1} - ${result.ignored ? 'ignored ' : ''}${
+				result.source
+			}`,
+		);
+
+		if (result.warnings.length > 0) {
+			lines.push('---', 'messages:');
```
Comment (Mouvedia) [inline lib/formatters/tapFormatter.js:17]: I am not sure but it seems to me that `\n` is missing.

> `YAMLBlock   := "  ---\n" (YAMLLine)* "  ...\n"`

File: lib/formatters/tapFormatter.js (hunk 4 above)
Comment (ybiquitous) [reply lib/formatters/tapFormatter.js:30]: We have implemented the version 13, not version 14:

https://github.com/stylelint/stylelint/blob/eb6d61357765f39ab7dceeb71d7749e55422564e/lib/formatters/tapFormatter.mjs#L7

But it seems v13 also requires an ending `...`, according to <https://testanything.org/tap-version-13-specification.html>:

> If the test line is immediately followed by an indented block beginning with `/^\s+---/` and ending with `/^\s+.../` that block will be interpreted as an inline YAML document.

So, it makes sense for me to fix `---` to `...`. Additionally, we might be able to update the TAP version from 13 to 14.

Can you create an issue or pull request if you have time?

File: lib/formatters/tapFormatter.js (hunk 5 above)
Comment (ybiquitous) [reply lib/formatters/tapFormatter.js:17]: See `return lines.join('\n');` at the end of the function.

File: lib/formatters/tapFormatter.js (hunk 4 above)
Comment (Mouvedia) [reply lib/formatters/tapFormatter.js:30]: I was reviewing which formatter would have the `fixed` mention.
see https://github.com/stylelint/stylelint/issues/4664#issuecomment-2159060803

> Can you create an issue or pull request if you have time?

Not right away no, but what I wonder is why it wasn't picked up for all these years.
The tap consumers must be very lax.
This is weird.

Comment (Mouvedia) [reply lib/formatters/tapFormatter.js:30]: Actually, I am planning to send a PR this weekend.

**Pull request merged**